server/src/bundle-validation.ts

```diff
@@ -0,0 +1,198 @@
+import { Connection, Diagnostic, DiagnosticSeverity } from 'vscode-languageserver/node'
+import { BookBundle, BundleItem } from './book-bundle'
+import { calculateElementPositions, expect, generateDiagnostic } from './utils'
+
+enum DiagnosticSource {
+  ImagePath = 'Image validation',
+  Link = 'Link validation'
+}
+
+export interface BundleValidationRequest {
+  causeUri: string
+}
+
+export class BundleValidationQueue {
+  private queue: BundleItem[] = []
+  private timer: NodeJS.Immediate | undefined
+  errorEncountered: Error | undefined
+
+  constructor(private readonly bundle: BookBundle, private readonly connection: Connection) {}
+
+  private clearQueue(): void {
+    this.queue = []
+  }
+
+  addRequest(request?: BundleValidationRequest): void {
+    this.clearQueue()
+    if (request != null) {
+      const priorityItem = expect(this.bundle.bundleItemFromUri(request.causeUri), 'caller must verify uri resides in this bundle')
+      if (priorityItem.type === 'collections' || priorityItem.type === 'modules') {
```

Comment:
I would think so. If a media file is added or removed, then we should re-check all modules (since they may or may not have a reference to that media file). Now, collections on the other hand, we probably don't need to re-queue on a media event. We could add a special case for that if we like.

Reply:
Ah, my mistake, my 👀  misread this somehow when I made the comment as I thought the full sweep wouldn't occur, but it does. 

Ignore the comment. I don't think adding the special case to exclude collections matters too much.